quic/s2n-quic-transport/src/dc/manager.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    contexts::WriteContext,
+    endpoint,
+    sync::{flag, flag::Writer},
+};
+use s2n_quic_core::{
+    ack,
+    crypto::tls,
+    dc,
+    dc::{Endpoint, Path},
+    ensure, event,
+    event::builder::{DcState, DcStateChanged},
+    frame::DcStatelessResetTokens,
+    packet::number::PacketNumber,
+    state::is,
+    stateless_reset, transmission,
+    transmission::interest::Query,
+};
+
+/// Manages transmission and receipt of `DC_STATELESS_RESET_TOKENS` and
+/// notifications to the dc::Endpoint Path
+pub struct Manager<Config: endpoint::Config> {
+    path: Option<<<Config as endpoint::Config>::DcEndpoint as Endpoint>::Path>,
+    version: Option<dc::Version>,
+    state: State,
+}
+
+type Flag = flag::Flag<DcStatelessResetTokenWriter>;
+
+#[derive(Debug, Default)]
+enum State {
+    #[default]
+    Init,
+    PathSecretsReady(Flag),
+    Complete,
+}
+
+impl State {
+    is!(is_init, Init);
+    is!(is_complete, Complete);
+    fn is_path_secrets_ready(&self) -> bool {
+        matches!(self, Self::PathSecretsReady(_))
+    }
+}
+
+impl<Config: endpoint::Config> Manager<Config> {
+    /// Constructs a new `dc::Manager` with the optional given path
+    ///
+    /// If path is `None`, the `dc::Manager` will be disabled
+    pub fn new<Pub: event::ConnectionPublisher>(
+        path: Option<<<Config as endpoint::Config>::DcEndpoint as Endpoint>::Path>,
+        version: dc::Version,
+        publisher: &mut Pub,
+    ) -> Self {
+        if path.is_some() {
+            publisher.on_dc_state_changed(DcStateChanged {
+                state: DcState::VersionNegotiated { version },
+            });
+            Self {
+                path,
+                version: Some(version),
+                state: State::Init,
+            }
+        } else {
+            Self::disabled()
+        }
+    }
+
+    /// Returns a disabled `dc::Manager`
+    pub fn disabled() -> Self {
+        Self {
+            path: None,
+            version: None,
+            state: State::Complete,
+        }
+    }
+
+    /// The dc version that was negotiated, if any
+    ///
+    /// Returns `None` if no version was negotiated or the `dc::Endpoint` did
+    /// not initialize a path for the connection
+    pub fn version(&self) -> Option<dc::Version> {
+        self.version
+    }
+
+    /// Called when the TLS session has indicated path secrets are ready
+    /// to be derived for the dc path
+    ///
+    /// Initiates sending of the `DC_STATELESS_RESET_TOKENS` frame
+    pub fn on_path_secrets_ready<Pub: event::ConnectionPublisher>(
+        &mut self,
+        session: &impl tls::TlsSession,
+        publisher: &mut Pub,
+    ) {
+        ensure!(!self.state.is_complete());
```

Comment:
should we also `ensure!(self.path.is_some())`? otherwise, we may be emitting an event that doesn't really make sense for the configured environment.

Reply:
if path is none, the state moves straight to Complete

Reply:
ah good call

Reply:
ended up getting rid of all the ensures and just using the states